File: Data Structures & Algorithms/max-water-container/submission-1.java
class Solution {
    public int maxArea(int[] heights) {
        int n = heights.length;
        int i = 0; 
        int j = n - 1;

        int max = 0;
        while(i < j) {
            int height = Math.min(heights[i], heights[j]);
            int width = j - i;
            max = Math.max(max, height*width);
            if(heights[i] < heights[j]) i++;
            else j--;
        }
        return max;
    }
}

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-1.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        n = len(heights)
        stack = []  # will contain [height, index]
        max_area = 0

        for i, h in enumerate(heights):
            start_index = i
            while stack and stack[-1][0] > h:
                height, index = stack.pop()
                area = height * (i - index)
                max_area = max(max_area, area)
                start_index = index     # because we can push the index backwards
            stack.append([h, start_index])
        
        while stack:
            height, index = stack.pop()
            area = height * (n - index)
            max_area = max(max_area, area)
        
        return max_area
        
